_posts/2019-05-26-redundant-order-by.md

```diff
@@ -0,0 +1,128 @@
+---
+layout: post
+title:  "Removing redundant ORDER BY"
+author: Martin Traverso
+---
+
+Sorting data in Presto is expensive. It's an operation that requires considerable amount of CPU and memory,
+and requires coordination that results in reduced parallelism. During query planning, Presto will insert
+a sorting step for each `ORDER BY` clause in a query, even though the construct has limited effect
+in the SQL language. This, combined with users and tools inadvertently using `ORDER BY` in places that
+have no effect, can result in severe performance degradation and waste of resources. We finally
+fixed this in [Presto 312](https://prestosql.io/docs/current/release/release-312.html)!   
+
+Quoting from the SQL specification (ISO 9075 Part 2):
+
+> A `<query expression>` can contain an optional `<order by clause>`. The ordering of the rows of the table
+ specified by the `<query expression>` is guaranteed only for the `<query expression>` that immediately 
+ contains the `<order by clause>`.
+
+This means, a query engine is free to ignore any `ORDER BY` clause that doesn't fit that context. Let's consider
+some examples where the clause is irrelevant.
+
+```sql
+INSERT INTO some_table 
+SELECT * FROM another_table 
+ORDER BY field 
+```
+
+While this query has the semblance of creating a sorted table, that's not so. Tables in SQL are inherently
+unordered. Once the data is written, there's no guarantee it will come out sorted when read. This is 
+particularly true for a parallel, distributed query engine like Presto that reads and processes data using
+many threads simultaneously. Note that some storage engines may store data sorted, but that is not controlled
+during data insertion. Executing the `ORDER BY` just causes the query to perform poorly due to reduced 
```

Comment:
It would be good to mention how to properly create a sorted table in eg Hive

Reply:
There is no way to create a globally sorted *table* in Hive, afaik. You can only insert data globally sorted using CLUSTER BY, but that's query by query. The only sorting you can describe for a table is per bucket.

Reply:
It might still be worth mentioning.Up to  you